api/src/main/java/org/apache/iceberg/metrics/DefaultTimer.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.metrics;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.StringJoiner;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.base.Stopwatch;
+
+/**
+ * A default {@link Timer} implementation that uses a {@link Stopwatch} instance internally to measure time.
+ */
+public class DefaultTimer implements Timer {
+  private final TimeUnit defaultTimeUnit;
+  private final AtomicLong count = new AtomicLong();
+  private Duration duration = Duration.ZERO;
+  private final AtomicReference<Stopwatch> current = new AtomicReference<>();
+
+  public DefaultTimer(TimeUnit timeUnit) {
+    Preconditions.checkArgument(null != timeUnit, "TimeUnit must be non-null");
+    this.defaultTimeUnit = timeUnit;
+  }
+
+  @Override
+  public long count() {
+    return count.get();
+  }
+
+  @Override
+  public Duration totalDuration() {
+    return duration;
+  }
+
+  @Override
+  public void startTimer() {
+    boolean updated = current.compareAndSet(null, Stopwatch.createStarted());
+    Preconditions.checkState(updated, "startTimer() called multiple times");
+  }
+
+  @Override
+  public void stopTimer() {
+    Stopwatch stopwatch = current.getAndSet(null);
+    Preconditions.checkState(null != stopwatch, "startTimer() was not called");
+    record(stopwatch.stop().elapsed(defaultTimeUnit), defaultTimeUnit);
+  }
+
+  @Override
+  public void record(long amount, TimeUnit unit) {
+    if (amount >= 0) {
+      duration = duration.plus(amount, toChronoUnit(unit));
+      count.incrementAndGet();
+    }
+  }
+
+  @Override
+  public <T> T record(Supplier<T> supplier) {
+    startTimer();
+    try {
+      return supplier.get();
+    } finally {
+      stopTimer();
+    }
+  }
+
+  @Override
+  public <T> T recordCallable(Callable<T> callable) throws Exception {
+    startTimer();
+    try {
+      return callable.call();
+    } finally {
+      stopTimer();
+    }
+  }
+
+  @Override
+  public void record(Runnable runnable) {
+    startTimer();
+    try {
+      runnable.run();
+    } finally {
+      stopTimer();
+    }
+  }
+
+  @Override
+  public String toString() {
+    return new StringJoiner(", ", DefaultTimer.class.getSimpleName() + "[", "]")
+        .add("count=" + count)
+        .add("duration=" + duration)
+        .toString();
+  }
```

Comment:
Style / Non-blocking: I'm a big fan of using Guava's `MoreObjects.toStringHelper` like seen here:

https://github.com/apache/iceberg/blob/90fe0edf1a671095e587a53adeb31cc84d01fb89/core/src/main/java/org/apache/iceberg/rest/responses/LoadTableResponse.java#L72-L78

It more or less gives you the same result but is a lot less mental overhead in my opinion. Up to you if you use it or not though.

Reply:
agreed, that looks much nicer